Python: update script for packages

FRidh opened the pull request:
This script determines whether there is an update available for the
package and if so, updates the version and sha256 attributes.

This only works for packages that have their own expression, and `pname`
and `sha256` attribute.

###### Motivation for this change


###### Things done

- [x] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] Linux
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

FRidh committed:
Python: update script for packages

This script determines whether there is an update available for the
package and if so, updates the version and sha256 attributes.

This only works for packages that have their own expression, and `pname`
and `sha256` attribute.

FRidh commented:
to do:

- [ ] the script may downgrade a package. Compare versions
- [ ] maybe clean up `extract_relevant_nix_data` which I recycled from elsewhere

FRidh commented:
Give it a try on the `python-modules` folder and see how many outdated packages there are already there.

Mic92 commented on maintainers/scripts/update-python-libraries:
```diff
@@ -0,0 +1,221 @@
+    else:
+        logging.warning("Request for {} failed".format(url))
+
+def _get_latest_version(package, extension):
```
future: github tags via API: https://api.github.com/repos/Mic92/python-mpd2/tags\?limit\=1

FRidh replied:
The GitHub API is indeed also very useful.

I've been thinking of putting together a repo where we list for each package where sources can be found. Something like
```
{ "numpy" :  [
  ("GitHub", { repo = ...; owner =... ; tagRegex = "..."; }),
  ("PyPI", { name = "numpy"}),
], "scipy" : [ ...
],
}
```
Then we can use the different fetchers to build up a list of available versions and files.

Or, we don't make such a list, and just pull everything from each and try to put it together, but then you end up with https://libraries.io. I think that's a nice project, although, as with all projects that pull data and interpret it, it makes mistakes.

lsix replied:
The `extension` parameter do not seem to be used. Could it be dropped?

FRidh replied:
For correctness I think we should use the extension, but my recycled `extract_relevant_nix_data` needs some work then, though not much.

Mic92 commented on maintainers/scripts/update-python-libraries:
```diff
@@ -0,0 +1,221 @@
+    """Extract relevant Nix data from the JSON of a package obtained from PyPI.
+
+    :param json: JSON obtained from PyPI
+    :param version: Specific version of package or one of the following strings: ['release',]
```
non-existing parameter.

Mic92 commented on maintainers/scripts/update-python-libraries:
```diff
@@ -0,0 +1,221 @@
+    summary = summary if summary != "UNKNOWN" else None
+    #description = str(json['info'].get('description'))
+    #description = description if description != "UNKNOWN" else None
+    homepage = json['info'].get('home_page')
```
Is this field mandatory on Pypi?

FRidh replied:
Its not a mandatory field, that's why I use `.get`. Note that this piece isn't used currently.

Mic92 commented on maintainers/scripts/update-python-libraries:
```diff
@@ -0,0 +1,221 @@
+    regex = re.compile(regex)
+    value = regex.findall(text)
+
+    return value[0]
```
Raise an descriptive error message, if a field was not found or more then once.

Mic92 commented on maintainers/scripts/update-python-libraries:
```diff
@@ -0,0 +1,221 @@
+        return False
+
+    with open(path, 'r') as f:
+        text = f.read()
```
I would strip `#.*` here to ignore comments

lsix replied:
Why would we do that? The purpose here is to only update the version / sha256 of the derivations definition. I think comments from the author are more than welcome (some may tell why tests are skipped).

Mic92 replied:
to avoid things like this when parsing:

```
# FIXME: bla
#version = "1.2.3";
version = 1.2.2;
```

I did not meant they should be stripped from the original file.

lsix replied:
If you want to avoid such situations, I would prefer that the matching regex looks more like:

```python
'^\s*{}\s+=\s+"(.*)";'.format(attribute)
```

dezgeg replied:
You could avoid parsing with calling something like `nix-instantiate --eval --strict --json -A ...`

FRidh replied:
Then I get names but I won't know where the packages reside nor whether I can actually replace the fields.

FRidh commented:
Thanks for your review @Mic92

lsix commented on maintainers/scripts/update-python-libraries:
```diff
@@ -0,0 +1,221 @@
+#! /usr/bin/env nix-shell
+#! nix-shell -i python3 -p 'python3.withPackages(ps: with ps; [ requests toolz unidecode ])'
```
`unidecode` do not seem to be used.

lsix commented on maintainers/scripts/update-python-libraries:
```diff
@@ -0,0 +1,221 @@
+    regex = '{}\s+=\s+"(.*)";'.format(attribute)
+    regex = re.compile(regex)
+    value = regex.findall(text)
+    if len(value):
```
I would say that if `len(value) != 1` it does not seem safe to be safe. If we find many matches (which should probably not occur in a regular build function), how would we say the first one is the good one ? I think I would prefer to raise an error in such situation.

lsix commented on maintainers/scripts/update-python-libraries:
```diff
@@ -0,0 +1,221 @@
+    else:
+        url = _get_value("url", text)
+        if url:
+            extension = os.path.splitext(url)[1]
```
Since the `extension` parameter do not seem to be used in the `_get_latest_version` function, computing it is not worthed. Also, this would miss the `.tar.*` like extensions.

dezgeg commented on maintainers/scripts/update-python-libraries:
```diff
@@ -0,0 +1,221 @@
+    if new_version != version:
+
+        text = _replace_value('version', new_version, text)
+        text = _replace_value('sha256', new_sha256, text)
```
Did you notice `./pkgs/common-updater/scripts/update-source-version` in the repo?

FRidh replied:
I wasn't aware of it, and it looks like a good building block. It seems to require the `meta.position` attribute that the common-updater requires. For the use case here I think the regex in combination with a target file/folder is simpler.

FRidh commented:
Thanks for the feedback. I've addressed most of the comments. Now it uses a lot of exception catching, not very pretty. I'm merging this because its already a time-saver the way it is now.
